src/wallet/rpcwallet.cpp

```diff
@@ -1720,6 +1739,47 @@ UniValue walletpassphrasechange(const UniValue& params, bool fHelp)
     return NullUniValue;
 }
 
+UniValue walletconfirmbackup(const UniValue& params, bool fHelp)
+{
+    if (!EnsureWalletIsAvailable(fHelp))
+        return NullUniValue;
+
+    if (fHelp || params.size() != 1)
+        throw runtime_error(
+            "walletconfirmbackup \"emergency recovery phrase\"\n"
```

Comment:
I am **VERY** uncomfortable with an RPC method that recommends passing the wallet's full recovery phrase on the CLI. This is almost certainly going to cause the mnemonic to be cached in the user's shell history file in plaintext. (A wallet GUI interacting with the RPC via HTTP doesn't have this issue, but users can and do interact via the CLI.)

There's a `-stdin` argument for `zcash-cli` that can be used to avoid this, and at a _MINIMUM_ we should only be using that in the helptext. I note that we should also alter `bitcoin-cli.cpp` to add `walletconfirmbackup` to the whitelist of "allowed hidden inputs", so that the text isn't echoed to the terminal as it is typed (like a passphrase).

However, I question whether this is the best RPC method design, or whether we should have a full challenge-response cycle (checking individual words in the phrase, either in-order or randomised order), that ensures the user isn't passing the entire phrase over in a single call. Typing in the entire mnemonic correctly via `-stdin` is also somewhat unlikely for the user to get right (a wallet GUI like ZecWallet obviously won't have this issue).

Ideally the RPC we end up adding would also be workshopped with the GUI that will be using it. I don't mind merging this PR before that happens, but we should get feedback from @adityapk00 before cutting the next release.

Reply:
Yes, typing in the phrase correctly would be error-prone, maybe just do something like this:
```bash
$ zcash-cli z_exportwallet mywallet
$ zcash-cli -stdin walletconfirmbackup < mywallet
```
`zcashd` looks through this big string for `recovery_phrase="word word word ..."` 

A problem may be if the string (the entire file) is too large. It may seem weird to send all this unnecessary data to `zcashd`. Maybe instead add a `zcash-cli` option that causes it to read stdin and filter just that line, and send the word list as the argument to the real RPC. In that case, it could take a file name. So it could become something like (note the dash argument means it's handled by `zcash-cli` locally).
```bash
$ zcash-cli z_exportwallet mywallet
$ zcash-cli -walletconfirmbackup mywallet
```
Maybe we could make it a bit forgiving too, so it would accept either a real wallet dump file, looking for a line like this:
```
# - recovery_phrase="open ceiling tobacco achieve search leopard hover solid buddy change next change into peanut average end text fiber allow grant answer nasty trigger yard"
```
Or it would accept a small file with just:
```
open ceiling tobacco achieve search leopard hover solid buddy change next change into peanut average end text fiber allow grant answer nasty trigger yard
```
If not too much effort, we could also allow the words to be on separate lines.

Reply:
In standup I had an idea for how to improve this: we add a `walletgetbackupchallenge` RPC method that returns an array of integers, corresponding to the order that the words in the mnemonic should be provided in `walletconfirmbackup`. This means that someone can't simply copy-paste the mnemonic phrase into the RPC; they need to actively think about their backup (as they already require them to do with mobile wallet workflows).

Then we alter `zcash-cli` to intercept `walletconfirmbackup` and:
- Call `walletgetbackupchallenge`.
- Interactively ask the user for the words in that order.
- Pass the words in the user-provided order to `walletconfirmbackup`.
  - We should have `walletconfirmbackup` actually take an array of strings, instead of a single string.

A GUI wallet like ZecWallet Full Node would implement this same logic itself. Someone trying to use `zcash-cli` programmatically would not be able to use `walletconfirmbackup` (intentionally, it needs to be interacting with a real user), so they would need to either go through this workflow themselves once, or choose to disable backup confirmation for their node.

Reply:
Also, if someone called `walletgetbackupchallenge` when the mnemonic has already been confirmed, it could invalidate the confirmation, preventing further use of mnemonic-requiring RPC methods until the backup is reconfirmed via `walletconfirmbackup`. This introduces a potential RPC DoS attack if someone can cause that RPC to be called, but (a) if an adversary can cause RPC calls then you have way worse problems, and (b) this "fails closed" in the same way that the alert system does: it prevents operations until the user intervenes, rather than exposing information.

Reply:
I appreciate your concern about the mnemonic being in cleartext in the user's shell history, but the challenge approach you're suggesting seems *way* too complicated for something that we're going to require of every user who wants to use the wallet; it seems to me that the user would be much more likely to just disable the requirement than jump through a bunch of hoops like that.

Reply:
Also, given that this is an RPC interface (not specifically a CLI interface) it seems like a bad idea to me to build this around a highly interactive workflow that requires multiple calls and non-atomic state being maintained by the backend. That requires writing a whole bunch more data to the backing database, because there's always the possibility that the server crashes in the midst of someone going through their confirmation process.

Reply:
> it seems to me that the user would be much more likely to just disable the requirement than jump through a bunch of hoops like that.

This is a risk we already have by providing that config option, even for the current implementation. We also similarly run the risk that a wallet GUI just automates the confirmation process without actually performing a backup or involving the user (though that is more effort on their part for my proposal). The difference right now is that the most straightforward way for a CLI user to resolve the "I can't generate keys" issue is almost guaranteed to result in unintentional disclosure of their mnemonic:
- The reduced friction of having the argument to `walletconfirmbackup` be trivially copy-pasteable from the exported wallet file makes it less likely that the user will stop to think about the consequences of where they paste it.
- There are actually two separate disclosure pathways:
  - The mnemonic gets cached in the shell history, which provides a persistent vector readable by anyone with permissions to read the user's shell history file (which may be a larger group than those with permissions to read the user's Zcash datadir or the location they store their wallet backup).
  - The mnemonic is revealed as a CLI argument in the process table, which discloses it to anyone who is actively reading the process table at the time (making cross-process disclosure entirely feasible, even if the other process does not have filesystem permissions).
    - IIRC this disclosure pathway was precisely why upstream added `-stdin` in the first place.
- We can't force users to use `-stdin`; by the time our code is running and could enforce it, the damage has already been done. We could still detect this situation and abort loudly to the user, but what are the meaningful remediation steps they could take?
  - We know at this time that the user hasn't generated any keys from the mnemonic, so their funds are safe, but it's potentially unsafe to use this wallet any further.
  - We can't tell them "delete your `wallet.dat` and make a new one with a fresh mnemonic" because this could be an existing wallet that a mnemonic was added to, and this check would be occuring in `zcash-cli` which has no knowledge of the age of the wallet.
  - We could recommend they look at their shell history and delete the mnemonic. For useful details we'd need to do a bunch of research to figure out all the possible ways that all the potential shell environments users could be using might save their history.
  - There's nothing we can do to remediate disclosure via the process table after the fact. We can't recommend users scour their system for active malicious processes, because few if any of our users will actually have the expertise to act on it.

> Also, given that this is an RPC interface (not specifically a CLI interface) it seems like a bad idea to me to build this around a highly interactive workflow that requires multiple calls and non-atomic state being maintained by the backend.

My initial idea for the challenge workflow involved many more RPC round trips (having the user request a single integer from the `zcashd` node, and then returning the single corresponding word), which would have provided a better UX (the user would get confirmation of correctness for each word individually, making it easier to figure out backup mistakes). I decided that wasn't worth the communication complexity, and settled on a single round-trip.

> That requires writing a whole bunch more data to the backing database, because there's always the possibility that the server crashes in the midst of someone going through their confirmation process.

I don't have a concern with the challenge array of integers being stored in-memory, and then if `zcashd` crashes then the challenge is lost and the user needs to start the confirmation workflow again. If the `zcashd` node crashed in the midst of the few minutes the user is going through the confirmation process, the user has other problems (e.g. the wallet GUI itself is probably just going to crash). We could make it possible for `zcashd` to detect a stale challenge by providing a "challenge ID" (a random `u32` would suffice) alongside the array of integers, and then checking it when `walletconfirmbackup` is called (if `zcashd` restarted in between, it would not know the challenge ID), enabling separate errors to be returned for "challenge response invalid" and "challenge needs to be restarted".

Reply:
@nuttycom asked in standup if we could do the `zcash-cli` interactive part without adding the extra challenge RPC. tl;dr I think that would not be as good, but would be an improvement over the current PR.

---

If the RPC method itself still has the API "first argument is the entire mnemonic phrase" then someone reading the RPC docs may think they can use `zcash-cli` in that way. This would cause the exact same disclosure issues I raised above, but now it wouldn't even confirm the backup, because `zcash-cli` would be intercepting the call to instead use an interactive process! We'd instead have to detect this case (user tries to confirm the backup, but provides the full mnemonic on the CLI) and provide all the same remediation responses.

That being said, if the _node_ documentation then says "call `zcash-cli walletconfirmbackup` with no args, and follow the prompts", that is a UX improvement over the current PR. There's a slight issue in that `zcash-cli` can't know how many mnemonic words the `zcashd` expects without at least one extra RPC (and if we add the extra RPC to query that, then we may as well implement the challenge logic), but if we simply always use 24 words (256 bits of entropy), then we can just assume that problem away.

Reply:
So, I looked through the RPC methods and there are a number of them for which the wallet passphrase could be passed by the command-line, without use of `-stdin` being enforced, so I don't see that the current implementation is substantially different from this situation? 

One possibility that occurs to me that'd be a little simpler is to just drop to 128 bits of entropy by requesting that the user send the first 12 words, or every other word?